### ISSUE TITLE
ReactDialog doesn't work properly when change attributes

### ISSUE THREAD

**Issue by saltylassi** [eclipse-theia/theia#12164]
### Bug Description:

If I change attribute of element (style, tabindex, and many things) and render something new elements, It still remains

### Steps to Reproduce:

1. Make React Dialog, like this sample code :
```ts
  override render(): React.ReactNode {
    return <div style={{ width: '300px', height: '300px' }}>
      {this.page === 1 ? this.renderFirst() : this.renderSecond()}
      <button onClick={() => {
        if (this.page === 1) {
          this.page = 2;
        } else {
          this.page = 1;
        }
        this.update();
      }}>nextPage</button>
      <button onClick={() => {
        const target = document.getElementById('div2');
        if (!target) {
          return;
        }
        target.style.backgroundColor = target.style.backgroundColor === 'red' ? 'blue' : 'red';
        this.update();
      }}>changeColor</button>
    </div>;
  }

  renderFirst(): React.ReactNode {
    return <div style={{ width: '100%', height: '100%', display: 'flex' }}>
      <div id='div1' style={{ width: '50%', border: '1px solid black' }}>page1</div>
      <div id='div2' style={{ width: '50%', border: '1px solid black' }}>page1</div>
    </div>;
  }

  renderSecond(): React.ReactNode {
    return <div style={{ width: '100%', height: '100%', display: 'flex' }}>
      <div style={{ width: '50%', border: '1px solid black' }}>page2</div>
      <div style={{ width: '50%', border: '1px solid black' }}>page2</div>
    </div>;
  }
```
2. click changeColor button (or change backgroundColor with chrome devtools)
3. click nextPage button
4. new elements rendered, but backgroundColor is **still red**.

### Additional Information

- Operating System: Windows 10
- Theia Version: 1.32

**Comment by vince-fugnitto** [comment]
@saltylassi I'm not entirely convinced it is a bug with the framework, just how `react` works and the fact there is no `state` being used. Do you mind creating a commit on a branch from which we can test from? At the moment the snippet is not useable.

**Comment by saltylassi** [comment]
@vince-fugnitto 

sorry. my fault

I mean ReactDialog, not React Dialog


this maybe works :

**1. make sample.tsx and bind**
``` Javascript
import { inject, injectable, interfaces } from '@theia/core/shared/inversify';
import { DialogProps } from '@theia/core/lib/browser';
import { ReactDialog } from '@theia/core/lib/browser/dialogs/react-dialog';
import * as React from '@theia/core/shared/react';

@injectable()
export class SampleDialogProps extends DialogProps {

}

@injectable()
export class SampleDialog extends ReactDialog<string> {

  private page: number = 1;

  constructor(@inject(SampleDialogProps) protected override props: SampleDialogProps) {
    super({ ...props });
  }

  override render(): React.ReactNode {
    return <div style={{ width: '300px', height: '300px' }}>
      {this.page === 1 ? this.renderFirst() : this.renderSecond()}
      {this.renderNextPageBtn()}
      {this.renderChangeColorBtn()}
    </div>;
  }

  renderFirst(): React.ReactNode {
    return <div style={{ width: '100%', height: '100%', display: 'flex' }}>
      <div id='div1' style={{ width: '50%', border: '1px solid black' }}>div1</div>
      <div id='div2' style={{ width: '50%', border: '1px solid black' }}>div2</div>
    </div>;
  }

  renderSecond(): React.ReactNode {
    return <div style={{ width: '100%', height: '100%', display: 'flex' }}>
      <div id='div3' style={{ width: '50%', border: '1px solid black' }}>div3</div>
      <div id='div4' style={{ width: '50%', border: '1px solid black' }}>div4</div>
    </div>;
  }

  renderNextPageBtn(): React.ReactNode {
    return <button onClick={() => {
      if (this.page === 1) {
        this.page = 2;
      } else {
        this.page = 1;
      }
      this.update();
    }}>nextPage</button>;
  }

  renderChangeColorBtn(): React.ReactNode {
    return <button onClick={() => {
      const target = document.getElementById('div2');
      if (!target) {
        return;
      }
      target.style.backgroundColor = target.style.backgroundColor === 'red' ? 'blue' : 'red';
      this.update();
    }}>changeColor</button>;
  }

  get value(): string {
    return '1';
  }
}

export const bindSample = (bind: interfaces.Bind) => {
  bind(SampleDialogProps).toConstantValue({ title: 'Sample' });
  bind(SampleDialog).toSelf().inSingletonScope();
};

```


**2. open this dialog, and you can see this :**

![1](https://user-images.githubusercontent.com/77769802/217966118-923c3d1e-8f1f-4c04-b0b6-1f11f2c6aa69.png)

in chrome devTools:
![1-1](https://user-images.githubusercontent.com/77769802/217966187-b2452f99-2047-400d-a894-75755093b927.png)


**3. click changeColor button**
![2](https://user-images.githubusercontent.com/77769802/217966344-e0fe2dc8-ee76-457a-a399-0646894ffec8.png)

```backgroundColor``` of ```div2``` becomes ```red```. button changes bgColor, and call this.update.

in chrome devTools:
![2-1](https://user-images.githubusercontent.com/77769802/217966460-85bdcdf0-4c41-4644-963e-502f24c618c7.png)


and then, click nextPage button, this button changes ```this.page``` and call ```this.update``` (like setState)

**4. click nextPage button to render new elements, div3 and div4**
![3-1](https://user-images.githubusercontent.com/77769802/217967324-3ca80013-8400-4876-91e4-eacc9dc624bc.png)

here is a problem: I've changed ```backgroundColor``` of ```div2```, not ```div4```, but ```backgroundColor``` of ```div4``` is red.

in chrome devTools:

![3](https://user-images.githubusercontent.com/77769802/217967292-e4f0a2bd-a263-4e0d-9fe3-42f0b4a06cbe.png)

**Comment by colin-grant-work** [comment]
@saltylassi, I don't think the problem is Theia's implementation of the ReactDialog or ReactWidget. Mixing direct DOM manipulation like this:

```typescript
      const target = document.getElementById('div2');
      if (!target) {
        return;
      }
      target.style.backgroundColor = target.style.backgroundColor === 'red' ? 'blue' : 'red';
```
with React's mechanisms is bad practice and can lead to undefined behavior. Basically, React expects to be responsible for the management of the DOM, and it attempts to make minimal changes to the DOM to reconcile its model of the DOM with the output of the latest render cycle. But using direct DOM manipulations means that the _actual_ DOM is not in synch with React's DOM model: React doesn't know that you set an additional background color on that DOM node, so it doesn't 'fix' it to the state you expect. But it does adjust all of the other properties to match the expectations of the Render method: it updates the ID and the inner text.

Basically, you need to choose: do you want to manage the DOM, or do you want React to manage the DOM? Mixing the two will lead to unpredictable and, in many cases, undesirable behavior.
